### ISSUE TITLE
Unblock Read on Close

### ISSUE THREAD

**Issue by larsgk** [bugst/go-serial#13]
At least on linux I have the problem that my reading goroutine just hangs in blocking Read until the cable is pulled.  One small hack/fix that seemed to work - but should probably be done in a cleaner way was:

```
// Close the serial port
func (port *SerialPort) Close() error {
	syscall.SetNonblock(port.handle, true)  <----
	port.releaseExclusiveAccess()
	return syscall.Close(port.handle)
}
```

thoughts?

**Comment by larsgk** [comment]
hmm... on a few more tries this seems unstable.  The solution is probably something else that I can't figure out at the moment.  Have you considered a solution to soft-killingl the serial connection while reading (blocking) ?

**Comment by larsgk** [comment]
Maybe embedding something like tarm's non-blocking mode (with timeout): https://github.com/tarm/serial

**Comment by larsgk** [comment]
ok - looking at the doc for the Reader interface: https://golang.org/pkg/io/#Reader

"Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0. Callers should treat a return of 0 and nil as indicating that nothing happened; in particular it does not indicate EOF."

So - what is normally EOF on linux (a 0 bytes read) should be mapped to an error and 0 read should be possible when setting to non-blocking mode.  I just made a fork and will try to fiddle for windows and linux.

**Comment by cmaglie** [comment]
as the Reader doc says:

> Implementations of Read are discouraged from returning a zero byte count with a nil error, except when len(p) == 0

So I won't do that unless really needed.

Instead for the non-blocking problem, I think that the idiomatic way for doing it is using a goroutine + channel, somehing along the line of what suggested here:
https://groups.google.com/forum/#!topic/golang-nuts/LL9fvORhVS0

**Comment by larsgk** [comment]
mmm.. ok.

Is there another way to make the Read jump out of block - e.g. by doing some more magic in Close so it forces it to EOF, e.g?

**Comment by cmaglie** [comment]
Close() should make the Read() call to exit with an error. Hope that this is happening otherwise it should be fixed in some way :-)

**Comment by larsgk** [comment]
Windows works - linux doesn't (it seems).  The Close command is called - but it doesn't unblock the Read it seems

**Comment by larsgk** [comment]
After reading a lot of docs on serial comm on linux - and looking at a few other implementations, I'd say it should be fine to have it as a optional functionality to have a non-blocking / timeout based mode....*unless* you can find a way to actually jump out of the Read when Close is called.

See my code here: https://github.com/larsgk/go-webserial/blob/master/webapi.go#L91-L119

**Comment by larsgk** [comment]
HA!.. getting closer... muhahaha

THIS go project by @nqv works very well with linux (close and all):  https://github.com/goburrow/serial
...but not very well on windows (I don't get data back from my board at all)

.. while your project works well on windows and not so much on linux (the close part at least)

Maybe you should merge into a super solution covering both :)

**Comment by cmaglie** [comment]
> HA!.. getting closer... muhahaha

LOL, ok I'll take a look, but you have to wait a bit, since this is a side project I have only the weekend to make some progress :-)

**Comment by larsgk** [comment]
@cmaglie ok then :) - I made a fork of the other repo and will try to embed your functioning windows code into it (if you don't mind?).   I see that there are like 3-4 parallel projects doing cross platform serial comm in go - it would probably make sense to have 1 really good and solid one... I wonder if google could be convinced to embed the best from all into an x package or so (as it's probably a quite common case)

**Comment by cmaglie** [comment]
> I made a fork of the other repo and will try to embed your functioning windows code into it (if you don't mind)?

It's open source :-) until you give credit and follow the license I'm absolutely ok with that

> I see that there are like 3-4 parallel projects doing cross platform serial comm in go - it would probably make sense to have 1 really good and solid one...

I've seen almost 30-40... there is a lot of cross "copy" between them, difficult to say which one is better maintained.

Going back to the issue: I've tried the following test program:

```go
package main

import "fmt"
import "log"
import "go.bug.st/serial.v1"
import "time"

func main() {
        port, err := serial.Open("/dev/ttyACM0", &serial.Mode{})
        if err != nil {
                log.Fatal(err)
        }

        fmt.Println("Port opened")

        go func() {
                fmt.Println("goroutine: reading")
                buff := make([]byte, 1000)
                n, err := port.Read(buff)
                fmt.Printf("goroutine: reading completed n=%d err=%+v\n", n, err)
        }()

        time.Sleep(time.Second*1)

        if err := port.Close(); err != nil {
                fmt.Printf("close err=%+v\n", err)
        }
        fmt.Println("Port closed")

        time.Sleep(time.Second*5)
}
```

and indeed the read is blocked forever even if the port is closed, here the output:

```
Port opened
goroutine: reading
Port closed
```

as we can see the goroutine never reach the `read completed` part.

Following your suggestion I've looked at @nqv implementation but even if I put a `select` before the `read` the subroutine will block on the `select`. So basically we are moving the problem from `read` to `select`.

BTW this makes me think that we can exploit the `select`call  to unblock the `read` by using a unix-pipe, for example if we `select` on both the serial port and the pipe. This way when the `Close()` method should do something like: 1) close the serial port, 2) send a byte to the pipe (so all the select are unblocked), 3) close the pipe.

**Comment by cmaglie** [comment]
I've pushed the proposed solution here #14

Seems to work, but I haven't tested extensively...

**Comment by larsgk** [comment]
GREAT! - I'll try it out in the morning :)
